src/lib/io/sol-uart-impl-linux.c

```diff
@@ -151,14 +229,35 @@ sol_uart_open(const char *port_name, const struct sol_uart_config *config)
     }
     tcflush(uart->fd, TCIOFLUSH);
 
-    uart->async.rx_fd_handler = sol_fd_add(uart->fd,
-        FD_ERROR_FLAGS | SOL_FD_FLAGS_IN, uart_rx_callback, uart);
-    if (!uart->async.rx_fd_handler) {
-        SOL_ERR("Unable to add file descriptor to watch UART.");
-        goto fail;
+    if (config->rx_cb) {
+        size_t rx_size;
+        void *rx_buf;
+
+        uart->fd_handler = sol_fd_add(uart->fd,
+            FD_ERROR_FLAGS | SOL_FD_FLAGS_IN, uart_fd_handler_callback, uart);
+        if (!uart->fd_handler) {
+            SOL_ERR("Unable to add file descriptor to watch UART.");
+            goto fail;
+        }
+
+        uart->rx_cb = config->rx_cb;
+        rx_size = config->rx_size;
+        if (!rx_size)
+            rx_size = DEFAULT_BUFFER_SIZE;
+
+        rx_buf = calloc(1, rx_size);
```

Comment:
minor, but here you could use `malloc()` since the contents do not need to be zeroed.